### PR TITLE
Adds Graphics test project

### PR DESCRIPTION
**Pull request by cugone**
Adds a test project to the solution that allow for local graphics testing.

----
This project is used to test the graphics features of nas2d with access to an up-to-date version without having to switch repos or constantly update a submodule.

Currently the tests are visual and manually verify:
  - Image rendering
  - Point rendering
  - Rectangle rendering
  - Filled Rectangle rendering
  - Circle Rendering
----

Note: There are a couple changes, and one optional, that must be made locally as they are not part of the repo.
- Set the Debugging Working Directory to `$(OutDir)..\..\` for all configs and all platforms:
1. Right-Click on test-graphics project
2. Properties...
3. Debugging Tree
4. Set `Configuration` to `All Configurations`
5. Set `Platform` to `All Platforms`
4. Set Working Directory option to: `$(OutDir)..\..\`
- Set the Startup Projects to both `test` and `graphics-test`, this will allow F5 to work as expected.
1. Right-Click on the Solution in the Explorer
2. Properties...
3. Startup Project Tree
4. Select `Multiple startup projects`
5. Set `test` to Start
6. Set `graphics-test` to Start

The project is set up as the lowest dependent of the other two projects. Higher-level projects such as `test` will not build or run `graphics-test` if they are built or run individually.

However, if you build `graphics-test`, `nas2d-core` then `test` will build/run before `graphics-test`. If any of the dependent projects have *build* errors, `graphics-test` will not run. A failed test is not a build error.

### CODE REVIEW

**Comment by cugone** [comment]
The vcpkg SDL2 being broken is now a problem...

**Comment by DanRStevens** [comment]
I may need to check with Leeor on this one. I know he has separate repositories for manual tests. I'm not sure if this should be included here or there.

For automated tests, I would like them in the core repository. If there is a way to automate this please let me know.

Additionally, this would need to  pass all CI builds to be considered for inclusion here. That wouldn't be a problem for a standalone repo.

I would also be hesitant to include this if it required config changes to the main project. Config changes for the new test project are fine.

**Comment by cugone** [comment]
Other than adding the project to the solution, there are no required changes to any other project. The above mentioned changes to the output directory and setting the startup project are stored in a non-tracked `.vcxproj.user` file that is stored locally and not on the repo. Anyone cloning the project would have to make the changes.

**Comment by cugone** [comment]
It all works, now. :D

File: test-graphics/TestGraphics.cpp [hunk 1]
```diff
@@ -7,29 +7,26 @@ static std::mt19937 generator;
 static std::uniform_int_distribution<int> jitter_distribution(0, 64);
 
 TestGraphics::TestGraphics()
-: mFont("fonts/opensans-bold.ttf", 16)
-, mPlanet("planets/planet_b.png")
-, mLogo("sys/nas2d_logo.png")
+: mDxImage("Test_StbiAndDirectX.png")
+, mOglImage("Test_OpenGL.png")
 {}
```
Comment (DanRStevens) [inline test-graphics/TestGraphics.cpp]: The other projects uses mostly line trailing punctuation:
```cpp
TestGraphics::TestGraphics() :
  mDxImage("Test_StbiAndDirectX.png"),
  mOglImage("Test_OpenGL.png")
{}
```

**Comment by DanRStevens** [comment]
I was able to get this to compile on Linux in an ad hoc way:

GCC:
```
g++ *.cpp -I ../include/ -std=c++17 -L ../lib/ -lnas2d $(sdl2-config --static-libs) -lSDL2_mixer -lSDL2_image -lSDL2_ttf -lGLEW -lGL -lphysfs
```
Clang:
```
clang++ *.cpp -I ../include/ -std=c++17 -L ../lib/ -lnas2d $(sdl2-config --static-libs) -lSDL2_mixer -lSDL2_image -lSDL2_ttf -lGLEW -lGL -lphysfs
```

----

Unfortunately, the output from either compiler crashes with a core dump during startup.

Backtrace:
```
#0  0x000055555555f1d2 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_data (this=0x1) at /usr/include/c++/8/bits/basic_string.h:176
#1  0x0000555555566626 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::c_str (this=0x1) at /usr/include/c++/8/bits/basic_string.h:2291
#2  0x000055555557d06f in NAS2D::Filesystem::open (this=0x5555557d04b0, 
    filename=<error reading variable: Cannot access memory at address 0x9>) at src/Filesystem.cpp:187
#3  0x0000555555573a8f in NAS2D::RendererOpenGL::addCursor (this=0x5555557e6950, 
    filePath=<error reading variable: Cannot access memory at address 0x9>, cursorId=1431779902, 
    offx=0, offy=1436614272) at src/Renderer/RendererOpenGL.cpp:477
#4  0x000055555555fc2d in TestGraphics::initialize() ()
#5  0x0000555555586e93 in NAS2D::StateManager::setState (this=0x7fffffffdbf0, state=0x555555a1ad10)
    at src/StateManager.cpp:70
#6  0x00005555555694b4 in NAS2D::Game::go (this=0x7fffffffdc41, state=0x555555a1ad10)
    at src/Game.cpp:136
#7  0x000055555555ec7f in main ()
```

**Comment by DanRStevens** [comment]
After comparing the code to the backtrace, the result appear to be nonsensical. I tried again, after rebuilding NAS2D, and it seems to work now. It was probably a bad link, where a symbol had been changed between compiling the library and compiling the test project, which led to improper linking.

GCC:
```
g++ *.cpp -g -std=c++17 -I ../include/ -L ../lib/ -lnas2d $(sdl2-config --static-libs) -lSDL2_mixer -lSDL2_image -lSDL2_ttf -lGLEW -lGL -lphysfs
```

Clang:
```
clang++ *.cpp -g -std=c++17 -I ../include/ -L ../lib/ -lnas2d $(sdl2-config --static-libs) -lSDL2_mixer -lSDL2_image -lSDL2_ttf -lGLEW -lGL -lphysfs
```

**Comment by cugone** [comment]
To answer your previous question of "Can this be automated?" No. Testing graphical output of a dynamic game is impossible: https://stackoverflow.com/a/3223515/421178

If we were just creating a tool that made static images, sure, have the GPU test to see if every pixel of the output matches every pixel of a reference image.

**Comment by cugone** [comment]
Any update on the status of this?

File: test-graphics/TestGraphics.cpp [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+#include <functional>
+#include <random>
+
+static std::mt19937 generator;
+static std::uniform_int_distribution<int> jitter_distribution(0, 64);
+
+TestGraphics::TestGraphics() :
+ mDxImage("Test_StbiAndDirectX.png"),
+ mOglImage("Test_OpenGL.png")
+{}
```
[the remark this answers is not in the document]
Comment (cugone) [reply test-graphics/TestGraphics.cpp]: Remind me: no indentation or one tab?

[the remark this answers is not in the document]
Comment (DanRStevens) [reply test-graphics/TestGraphics.cpp]: One tab.

**Comment by cugone** [comment]
I've changed the test images to our own custom images that test the same features.

**Comment by DanRStevens** [comment]
Ok, I'm going to merge this now.

I would still like to resolve the startup working directory issue so local edits are not needed. Preferably I'd like to do that without custom project file modifications. We can figure something out post merge. I plan to do some repository re-organization, which may affect how we solve this, so I want to get that out of the way first.

We'll need Linux `makefile` support to compile this project from Linux.

We'll likely merge the nas2d-tests repo soon, so the test projects will need to be organized somehow.

**Comment by DanRStevens** [comment]
Actually, I was just thinking, what if we add a data path command line option. That would make things pretty flexible. That would make it pretty convenient to run from the `makefile`, and probably good practice to allow anyway.

That doesn't seem to solve the problem on Windows though. You could set a startup parameter, but I suspect that will also be in the user settings, and so not stored in the repository.

**Pull request merged**